### ISSUE TITLE
sx prop re-renders

### ISSUE THREAD

**Issue by nandorojo** [nandorojo/dripsy#12]
I really like the API of being able to pass styles directly to components inline using `sx`.

The one downside is that it can trigger re-renders on every parent render, since the styles create a new `sx` object prop.

A few possible solutions:

1. Follow a `styled` components type approach, and create a `styled` function:

```js
import { styled, Box } from 'dripsy'

const Page = styled(Box)(props => ({
  backgroundColor: ['primary', 'secondary']
}))
```

2. Could there be some sort of smart memoization / deep comparison that checks the `sx` prop, and doesn't re-render if it matches? I am usually very, very way of deep comparison, as it's not recommended by react in almost all cases (see [here](https://twitter.com/dan_abramov/status/1104414707082055680).) 

One option could be to JSON.stringify the `sx` prop and compare if this changed. The problem with that approach is that styles can a) contain animated values, and b) have functions:

```js
// not JSON serializable...
<View sx={{ height: theme => theme.sizes[1] }} />
```

Basically, it would be nice if there were no concerns about passing style props inline for performance. I haven't done extensive testing on this, but I imagine it could have some downsides.

**Comment by cmaycumber** [comment]
Not sure if this is 100% related but it does relate to the sx prop. Is jsx pragma something that we want to consider including?

Now that you brought this up I'm not sure what the performance implications might be or if it's even necessary if we export the majority of components, but I can see a scenario where someone might want to style a component outside of the library with the sx prop.

**Comment by nandorojo** [comment]
> Is jsx pragma something that we want to consider including?

I hadn't thought about it too much. I'm open to it, the only problem is that the pragma uses just CSS styles, whereas this lib requires hooks to track the screen width and such. I can see how it's really convenient. I personally prefer just importing components and using them, but if people find it really convenient maybe it'd be better to try to include it. We would have to wrap the entire component that is receiving the `sx` prop with the `createThemedComponent` function, rather than just inject a style function the way `theme-ui` does.

**Comment by nandorojo** [comment]
I thought more about the performance here. I think the easiest solution is to just use a `styled` function:

```jsx
import { styled, View } from 'dripsy'

const Background = styled(View)({
  backgroundColor: ['secondary', 'primary']
})

// or
const Background = styled(View)(props => ({
  backgroundColor: ['secondary', props.selected ? 'primary' : 'grey']
}))
```

And similar to `styled-components`, you'd declare these outside of your component code.

**Comment by cmaycumber** [comment]
> I hadn't thought about it too much. I'm open to it, the only problem is that the pragma uses just CSS styles, whereas this lib requires hooks to track the screen width and such. I can see how it's really convenient. I personally prefer just importing components and using them, but if people find it really convenient maybe it'd be better to try to include it. We would have to wrap the entire component that is receiving the sx prop with the createThemedComponent function, rather than just inject a style function the way theme-ui does.

I don't mind importing the component either. The main use case for the jsx pragma in my experience is the ability to style components from outside libraries without needing to create a new component. Obviously this is easily alleviated by using the styled function or createThemedComponent function to wrap a third party component but it can be handy to just style it using sx without anything but the jsx pragma. Personally I'm fine with either approach and don't have a huge opinion either way.

> I thought more about the performance here. I think the easiest solution is to just use a styled function:

Would this replace the sx prop entirely? I personally think the styled function is a less elegant solution than the sx prop because in my experience it often leads to writing much more code than you would with just the sx prop. It's also annoying to have to pass additional props to the styled component then redeclare the prop types for your newly styled component when using typescript. 

Performance is definitely important especially when we start to get on the native side of things. I'm curious if there's a way to get the sx prop without a drop in performance or unnecessary rerenders. Does this library [https://github.com/samjbmason/theme-ui-native](https://github.com/samjbmason/theme-ui-native) run into a similar issue? I'm not sure if they have something that we're missing. I'd be glad to help look into a solution for this because I do feel strongly about the sx prop.

**Comment by nandorojo** [comment]
I definitely view the `sx` prop as staying forever and being the primary API. The `styled` function would just be an option if you want to avoid re-renders / style your components outside of the scope. It would be strictly optional and secondary to the `sx` prop.

**Comment by nandorojo** [comment]
I'd love to look into performance optimizations for the `sx` prop too, since I'll likely rely on this API often as well.

**Comment by nandorojo** [comment]
This may be relevant https://github.com/FormidableLabs/react-fast-compare

**Comment by nandorojo** [comment]
Another idea thinking about this. Not the most "simple", but dripsy could implement something like https://github.com/andywer/use-inline-memo and provide a `useSx` hook that lets you generate memoized styles.

**Comment by nandorojo** [comment]
This article (https://sebastienlorber.com/records-and-tuples-for-react) by @slorber about upcoming Records/Tuples appears to be promising. I'm not too familiar with how they'll work yet, but this is my impression.

It seems like records would help for any SX prop that only uses primitive values within objects. For example:

```tsx
// works
const Memoized = () => <Box
  sx={#{ bg: #['blue', 'red'] }}
/>

// fails
const Fails = () => <Box
  sx={#{ bg: theme => [theme.primary, theme.secondary] }}
/>
```

For the case of wanting to use dynamic theme values, we could allow the `sx` prop to be a function, as discussed above, where it returns an immutable record:

```jsx
const Styled = () => <Box
  sx={theme => {
    return #{ bg: #[theme.primary, theme.secondary] }
  }}
/>

For the function prop example here, we could have a `React.memo` function that compares `prevProps.sx(theme) === nexxtProps.sx(theme)`, and not re-render if they match.
```

**Comment by nandorojo** [comment]
I'll close this once the `styled` issues are fixed, and when the style prop is moved to `StyleSheet.create()`

**Comment by cmaycumber** [comment]
What does it take to move the style prop over to ```Stylesheet.create()```. Is it a matter of simply wrapping the styles as they are being processed in the css function?

**Comment by nandorojo** [comment]
Yeah, I believe so. I think I'd turn them into stylesheet objects in `useMemo` in `createThemedComponent`. It's unclear to me how much better this would be, but I see on [RNW's docs](https://necolas.github.io/react-native-web/docs/?path=/docs/guides-style--page#defining-styles) that it's recommended for better performance. As for the `responsiveSSRStyles`, I'd probably have to loop through each one of these and use `StyleSheet.create` for each.

I think it would have no effect on native, where `StyleSheet.create` is nothing more than a JS object.

**Comment by nandorojo** [comment]
I'm bubbling this issue back up.

Lately I've been thinking about static extraction of styles.

[This tweet](https://twitter.com/natebirdman/status/1312070631593185280?s=21) prompted me to think about performance solutions.

Ideally, any `sx` prop that receives an object inline that is static (ie doesn't use any variables from render code), should get turned into an object outside of render code. This means that any static styles get generated at build time, not runtime. This prevents react from recalculating styles on the fly. We could also use `StyleSheet.create` outside of render code.

Perhaps this could be done as some sort of Babel/Webpack plugin. I wish I had experience with either, but I don't. It could fit nicely with the JSX pragma, perhaps(?)

Some inspiration for the idea:
- https://github.com/martinandert/babel-plugin-css-in-js
- https://github.com/z0al/react-native-styled.macro
- https://github.com/jsxstyle/jsxstyle/ (see webpack plugin) heard through

**Comment by nandorojo** [comment]
Could use this maybe: https://github.com/kentcdodds/babel-plugin-macros

**Comment by slorber** [comment]
For what it's worth there are some useful presets here
https://github.com/jamiebuilds/babel-react-optimize

But not sure it handles this case.

Read the babel handbook to understand how to write such optim plugin



Le lun. 2 nov. 2020 à 07:47, Fernando Rojo <notifications@github.com> a
écrit :

**Comment by nandorojo** [comment]
Got it, good to know, thanks @slorber. I'll look into that.

I got some guidance from it [here](https://github.com/z0al/react-native-styled.macro/issues/19). It seems that it might be a bit complicated, given that we have to traverse the root app theme to determine if styles are constants. Not really sure since I'm so new to Babel.

**Comment by nandorojo** [comment]
This project from @natew looks very interesting. He mentioned to me on Twitter that he plans to use the array syntax for media queries, too. Since snack UI figured out a way to compile down to HTML views, using it under the hood could mean abandoning Fresnel for an approach that is closer to the metal. This assumes Snack UI ends up figuring out how to do media queries with CSS, meaning it's compatible with SSR.

(It also assumes all styles map from react native primitives to web styles the way that RNW intends. I haven't tested, but it would appear that it does.)

Link: https://github.com/natew/snackui

**Comment by nandorojo** [comment]
More discussion on this at https://github.com/necolas/react-native-web/issues/1801 and https://github.com/react-native-community/discussions-and-proposals/issues/254#issuecomment-650392950.

**Comment by nandorojo** [comment]
I've been looking more at Snack UI: https://github.com/snackui/snackui

The static and Babel plugin truly go over my head, but it might be worth trying to learn how this all works. Seems like it makes pretty big performance strides.

**Comment by cmaycumber** [comment]
Yeah, it looks pretty cool. From my understanding, if we were to create a babel plugin, would we be able to drop fresnel in lieu of media queries?

**Comment by nandorojo** [comment]
Good question...I'm actually not sure. Judging by this plugin alone, I think the answer is no, since it seems like they use JS client-side media queries and thus don't support SSR.

If we built on top of it, possibly. 

It seems like a key benefit we'd get is that any `View`s that don't use RN logic would be compiled down to simple `div`s, and their styles would use optimized CSS rather than recomputing constantly. Apparently this leads to much snappier rendering.

As for getting rid of fresnel, this would require applying our styles in pure CSS. Maybe we could add this feature for the simple `div`s, but any RN `View` with more logic would probably still require fresnel. The only workaround would be to `patch-package` RNW to allow `className`, which doesn't seem like a great idea.

On another note, I have found the `Sizer` component I mentioned in another issue super useful in my app. I also creates a native `div` and uses `theme-ui` for breakpoints, meaning it only uses CSS. It's nice for flex boxes and such. I plan on adding it to Dripsy's core.

**Comment by nandorojo** [comment]
This is certainly a lower priority, but I might look into it as a weekend project.

**Comment by nandorojo** [comment]
Recent optimizations with stylesheet should solve most of this, since it's now cached and stringified.

**Comment by intergalacticspacehighway** [comment]
@nandorojo @cmaycumber  Do you think a plugin like this might be worth? 
```jsx
function App() {
  return (
    <>
      <View sx={{ backgroundColor: "black" }} />
      <View sx={{ marginTop: "100", paddingVertical: 20 }} />
    </>
  );
}
```

**Converts to**
```jsx
import { useMemo as _useMemo } from "react";

function App() {
  const _sx = _useMemo(
    () => ({
      1: {
        backgroundColor: "black",
      },
      2: {
        marginTop: "100",
        paddingVertical: 20,
      },
    }),
    []
  );

  return (
    <>
      <View sx={_sx["1"]} />
      <View sx={_sx["2"]} />
    </>
  );
}

```
## Cases
### 1. sx remains untouched if it contains any variables.

```jsx
function App() {
  const padding = 10;
  return (
    <>
      <View sx={{ backgroundColor: "black", padding }} />
      <View sx={{ marginTop: "100", paddingVertical: 20 }} />
    </>
  );
}
```

**will be converted to**
```jsx
import { useMemo as _useMemo } from "react";

function App() {
  const _sx = _useMemo(
    () => ({
      2: {
        marginTop: "100",
        paddingVertical: 20,
      },
    }),
    []
  );

  const padding = 10;
  return (
    <>
      <View
        sx={{
          backgroundColor: "black",
          padding,
        }}
      />
      <View sx={_sx["2"]} />
    </>
  );
}
```
This case is considered because when it contains a variable, it gets harder to find location to put useMemo.
e.g.
```jsx
function App() {
    if (boolean) return null;

    const a = 10;
    return <View  sx={{padding: a}} />
}
```

Here, we might have to move the variable `a` above the condition and add useMemo.

### 2. Nit pick but it's hard to tell if a function returning JSX will be used as a react component.

e.g.
```jsx
const Component = () => <View sx={{bg: "black" }} />
```

can be used as `<Component />` or `Component();`

This makes it difficult to recognize whether useMemo should be added or not, one heuristic I can think of is if a function name starts with a capital letter it can be considered as a react component and try to memoize sx props.

**Comment by nandorojo** [comment]
@intergalacticspacehighway very cool! what do you think of this case?

```tsx
const Component = () => {
  const renderBox = () => <View sx={{ bg: 'background' }} />

  return <View>{renderBox()}</View>
}
```

I also was thinking that under the hood, I could use the new [SWR hashing function](https://github.com/vercel/swr/blob/master/src/utils/hash.ts) to deep compare `sx` props under the hood, created from this issue: https://github.com/vercel/swr/pull/1429

**Comment by intergalacticspacehighway** [comment]
@nandorojo  This output was generated from the plugin. The plugin is simple, it works like this.

1. Verify function starts with capital letter (To detect react component)
2. Traverse all the JSX elements and search for sx attribute that is an object. e.g. sx={{x: y}}
3. Check if sx doesn't have a variable, If not, prepare to put it in useMemo or else don't touch it.
4. Construct the memo object and put it at the start of the function and replace sx attributes.

I can push a version and a playground file where one can test all the cases.

```jsx
import { useMemo as _useMemo } from "react";

const Component = () => {
  const _sx = _useMemo(() => ({
    1: {
      bg: 'background'
    }
  }), []);

  const renderBox = () => <View sx={_sx["1"]} />;

  return <View>{renderBox()}</View>;
};
```

The hashing function looks cool! Maybe we can test the performance with new hash function on any app and try the babel plugin as well and verify if it adds any overall benefit.

**Comment by nandorojo** [comment]
Technically, if there are no variables, the sx variable could be outside of render. However, this could make things more complicated if there are multiple components, etc

**Comment by intergalacticspacehighway** [comment]
Yes. This can be done by creating global like object and using it for each component with an identifier. I still think useMemo "might" be better as it'll be shortlived and gets garbage collected on component unmount. This needs performance testing to confirm though.

**Comment by nandorojo** [comment]
Yeah the garbage collection is a good point

**Comment by nandorojo** [comment]
Dripsy v3 will automatically memoize the `sx` prop under the hood using `stableHash` from `swr`. See [here](https://github.com/nandorojo/dripsy/pull/124/commits/b9a15464a380fdb2e0346a536fb448a8489b25df#diff-8ce561232f7fced53e331bcb44b040c8051bc86a90780b86793a261a6b328aaaR51)
